### ISSUE TITLE
index: Invoking PGP results in tty conflict

### ISSUE THREAD

**Issue by alejandro-colomar** [neomutt/neomutt#4426]
## NeoMutt Version

Any version, I think.  I also tried mutt(1), and has the same behavior.
Anyway:

```sh
$ neomutt -v | head
NeoMutt 20240425-77-707b65
Copyright (C) 2015-2024 Richard Russon and friends
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Linux 6.11.4-amd64 (x86_64)
ncurses: ncurses 6.5.20240427 (compiled with 6.5.20240427)
libidn2: 2.3.7 (compiled with 2.3.7)
GPGME: 1.18.0
```

## Expected Behaviour

When opening an encrypted message, neomutt(1) should call gpgme to decrypt it, which itself invokes gpg-agent(1) (probably), which itself invokes pinentry(1) (probably), which should allow me to enter my passphrase and decrypt the message.

## Actual Behaviour

My pinentry(1) is linked to pinentry-tty(1):

```
$ which pinentry | xargs realpath
/usr/bin/pinentry-tty
```

It seems to conflict with curses, which is used by neomutt(1), and results in a broken window.  It seems like a race for the tty between neomutt(1)'s curses and pinentry-tty(1).

Cc: @dd9jn 

I end up with the following broken screen:

```
---NeoMutt: ~/mail/migadu/foss/Inbox [Msgs:4413/4413 Old:3 142M]---(threads/last
Invoking PGP...Please enter the passphrase to unlock the OpenPGP secret key:
                                                                            "Alejandro Colomar <alx@alejandro-colomar.es>"
                                          4096-bit RSA key, ID 00C14A7DBBDD521C,
                                                                               created 2024-02-02 (main key ID D57633D441E25BB5).

                                                 Passphrase: 
```

I usually have to just close the terminal, open a new one, run gpg to unlock the keyring, and then run neomutt(1) with the keyring already unlocked which shows the contents without running pinentry(1).

## Steps to Reproduce

0) Install `pinentry-tty`.
1)  `$ neomutt;`
2)  In the index, find an encrypted message.
3)  Press enter, to open the message.
4)  Try to enter the passphrase, and you'll realize that you cant.

## How often does this happen?

- Always

## When did it start to happen?

- Always.  I was just lazy to debug it, and could live with the workaround for some time.

## Extra Info

* Operating System and its version

Debian Sid.

```
$ dpkg -l | grep pinentry
ii  pinentry-tty                                             1.2.1-4+b1                               amd64        minimal dumb-terminal PIN or pass-phrase entry for GnuPG
```

* Were you using multiple copies of NeoMutt at once?

No.

* Were you using 'screen' or 'tmux'?

No.

* Is your email local (maildir) or remote (IMAP)?

Local.  mbsync(1) FTW!  :)

**Comment by alejandro-colomar** [comment]
It might be a pinentry-tty(1) bug, since I would expect it to take over the terminal for security reasons.

@dd9jn

**Comment by dd9jn** [comment]
Hi!

> When opening an encrypted message, neomutt(1) should call gpgme to
> decrypt it, which itself invokes gpg-agent(1) (probably), which itself
> invokes pinentry(1) (probably), which should allow me to enter my
> passphrase and decrypt the message.

I can't remember the details but Mutt needs to call mutt_need_hard_redraw
after a signing or decryption operation if a tty or curses pinentry is
used.  This used to work but I didn't follow the (neo)mutt development
closely the last years.

I also had in mind that we had an optimization in gpgme which watched
out for the PINENTRY_LAUNCHED status code so that applications can check
a flag to see whether a redraw is required.  I can't find it right now,
maybe someone else can give a hint.  Might have been some test branch
we decided not to use.


Shalom-Salam,

   Werner

-- 
The pioneers of a warless world are the youth that
refuse military service.             - A. Einstein

**Comment by dd9jn** [comment]
Here is what I had in mind in my last mail:

  if ((s = gpgme_get_ctx_flag (ctx, "redraw")) && *s)
    fputs ("Screen redraw suggested\n", stdout);

(from gpgme/tests/run-sign.c).  Use mutt_need_hard_redraw () instead of
the fputs ;-)


Salam-Shalom,

   Werner

-- 
The pioneers of a warless world are the youth that
refuse military service.             - A. Einstein

**Comment by alejandro-colomar** [comment]
Hi @dd9jn !

> Here is what I had in mind in my last mail:
> 
>   if ((s = gpgme_get_ctx_flag (ctx, "redraw")) && *s)
>     fputs ("Screen redraw suggested\n", stdout);
> 
> (from gpgme/tests/run-sign.c).  Use mutt_need_hard_redraw () instead of
> the fputs ;-)
> 
> 
> Salam-Shalom,
> 
>    Werner
> […](#)

Sorry for being slow to respond.  I was busy with other stuff.

neomutt(1) is not using fputs(3) there.  See:

```c
$ grep -rn gpgme_get_ctx_flag -C5
ncrypt/crypt_gpgme.c-118- * redraw_if_needed - Accommodate for a redraw if needed
ncrypt/crypt_gpgme.c-119- * @param ctx GPGME handle
ncrypt/crypt_gpgme.c-120- */
ncrypt/crypt_gpgme.c-121-static void redraw_if_needed(gpgme_ctx_t ctx)
ncrypt/crypt_gpgme.c-122-{
ncrypt/crypt_gpgme.c:123:  const char *s = gpgme_get_ctx_flag(ctx, "redraw");
ncrypt/crypt_gpgme.c-124-  if (!s /* flag not known */ || *s /* flag true */)
ncrypt/crypt_gpgme.c-125-  {
ncrypt/crypt_gpgme.c-126-    mutt_need_hard_redraw();
ncrypt/crypt_gpgme.c-127-  }
ncrypt/crypt_gpgme.c-128-}
```

**Comment by alejandro-colomar** [comment]
```sh
$ git blame -- ncrypt/crypt_gpgme.c | grep -C5 gpgme_get_ctx_flag
7a41cd9370 ncrypt/crypt_gpgme.c (Richard Russon (@flatcap)                 2021-10-20 18:27:19 +0100  118)  * redraw_if_needed - Accommodate for a redraw if needed
25bff4295d ncrypt/crypt_gpgme.c (Richard Russon (@flatcap)                 2018-07-15 23:33:48 +0100  119)  * @param ctx GPGME handle
5b823ab136 ncrypt/crypt_gpgme.c (Richard Russon (@flatcap)                 2017-07-16 20:49:10 +0100  120)  */
41b5f7640f crypt_gpgme.c        (Richard Russon (@flatcap)                 2017-04-12 19:45:24 +0100  121) static void redraw_if_needed(gpgme_ctx_t ctx)
1689cb1e48 crypt-gpgme.c        (Pietro Cerutti (@gahr)                    2017-02-01 15:52:40 +0000  122) {
41b5f7640f crypt_gpgme.c        (Richard Russon (@flatcap)                 2017-04-12 19:45:24 +0100  123)   const char *s = gpgme_get_ctx_flag(ctx, "redraw");
3ff204ca3e ncrypt/crypt_gpgme.c (Richard Russon (@flatcap)                 2017-05-23 14:13:41 +0100  124)   if (!s /* flag not known */ || *s /* flag true */)
1689cb1e48 crypt-gpgme.c        (Pietro Cerutti (@gahr)                    2017-02-01 15:52:40 +0000  125)   {
41b5f7640f crypt_gpgme.c        (Richard Russon (@flatcap)                 2017-04-12 19:45:24 +0100  126)     mutt_need_hard_redraw();
1689cb1e48 crypt-gpgme.c        (Pietro Cerutti (@gahr)                    2017-02-01 15:52:40 +0000  127)   }
3b0cec02a7 ncrypt/crypt_gpgme.c (UPSTREAM Werner Koch                      2018-12-03 08:41:55 +0100  128) }
```

**Comment by alejandro-colomar** [comment]
Ping @dd9jn

**Comment by dd9jn** [comment]
Okay, qhat is the question?  You changed the logic in assumption that some very old gpgme does not support the flag.  I don't think this is needed .   
(fputs was just an example).

**Comment by alejandro-colomar** [comment]
> Okay, qhat is the question? You changed the logic in assumption that some very old gpgme does not support the flag. I don't think this is needed . (fputs was just an example).

I don't really understand that code.  What would you suggest using there?

**Comment by dd9jn** [comment]
gpgme_get_ctx_flag returns whether you should redraw the screen.  It
can decide this by looking at status messages from the Pinentry.  Thus
only if a pinentry popped up the function will return true and you
should redraw.  The code above is correct; it does a redraw only when needed.

True is a defined as: For a boolean flag an empty string is returned
for False and the string "1" for True; thus either atoi or a simple
string test can be used to get the bolean value.

```
  const char *s = gpgme_get_ctx_flag(ctx, "redraw");
  if (!s /* flag not known */ || *s /* flag true */)
      mutt_need_hard_redraw();
```

Th function returns either a string or NULL which is a assigned top S.
The first test check whether NULL is returned which means the
requested flag is unknown.  Now that we know we have not NULL for S we
can dereference is (*s).  This is the same as if we would do

```
  if (s)
    {
       if (s[0])
         {
            mutt_need_hard_redraw ();
         }
    }
```


An empty string will have a value of zero for s[0] (aka *s) which
means False.  Note also that an explict test like s == NULL is not
required in C because the languages defines that in a condition a NULL
pointer has always the value False; regardless of its internal
representation (i.e. segmented memory models like in old style DOS
programs)

**Comment by alejandro-colomar** [comment]
> gpgme_get_ctx_flag returns whether you should redraw the screen. It can decide this by looking at status messages from the Pinentry. Thus only if a pinentry popped up the function will return true and you should redraw. The code above is correct; it does a redraw only when needed.

But the behavior isn't correct.  I can reproduce the bug with this code.

**Comment by alejandro-colomar** [comment]
> gpgme_get_ctx_flag returns whether you should redraw the screen. It can decide this by looking at status messages from the Pinentry. Thus only if a pinentry popped up the function will return true and you should redraw. The code above is correct; it does a redraw only when needed.
> 
> True is a defined as: For a boolean flag an empty string is returned for False and the string "1" for True; thus either atoi or a simple string test can be used to get the bolean value.
> 
> ```
>   const char *s = gpgme_get_ctx_flag(ctx, "redraw");
>   if (!s /* flag not known */ || *s /* flag true */)
>       mutt_need_hard_redraw();
> ```
> 
> Th function returns either a string or NULL which is a assigned top S. The first test check whether NULL is returned which means the requested flag is unknown. Now that we know we have not NULL for S we can dereference is (*s). This is the same as if we would do
> 
> ```
>   if (s)
>     {
>        if (s[0])
>          {
>             mutt_need_hard_redraw ();
>          }
>     }
> ```

BTW, this doesn't seem exactly the same.  The same would be:

```c
if (s) {
    if (s[0])
        mutt_need_hard_redraw();
} else {
    mutt_need_hard_redraw();
}
```

**Comment by dd9jn** [comment]
You are right, I had my version  w/o the fallback for very old gpgme versions in min (s && *s).

The problem at hand might be that the pinentry-tty version does not properly send the PINENTRY_LAUNCHED status lines.  Try with a standard pinentry or put "debug-pinentry" and "log-file /some/file" into gpg-agent.conf to see which status lines you get.

**Comment by alejandro-colomar** [comment]
I don't get anything in the log.

```sh
alx@devuan:~/.gnupg$ ls -l gpg-agent.conf 
-rw------- 1 alx alx 53 Mar 31 14:35 gpg-agent.conf
alx@devuan:~/.gnupg$ cat gpg-agent.conf 
debug-pinentry
log-file /home/alx/.gnupg/alx-dbg.log
alx@devuan:~/.gnupg$ ls /home/alx/.gnupg/alx-dbg.log
ls: cannot access '/home/alx/.gnupg/alx-dbg.log': No such file or directory
alx@devuan:~/.gnupg$ xfce4-terminal -e bash
alx@devuan:~/.gnupg$ ls /home/alx/.gnupg/alx-dbg.log
ls: cannot access '/home/alx/.gnupg/alx-dbg.log': No such file or directory
```

Within the `xfce4-terminal -e bash` is where I run neomutt(1) and try to decrypt a message, and I get the following screen:

```
6622   C Nov 04 Werner Koch     (3.7K) ┬─>Re: [neomutt/neomutt] index: Invoking
6623   C Nov 04 Werner Koch     (2.5K) ├─>Re: [neomutt/neomutt] index: Invoking
6624   C Mar 31 Werner Koch     (4.2K) ├─>Re: [neomutt/neomutt] index: Invoking
6625   C Mar 31 Werner Koch     (8.1K) ├─>Re: [neomutt/neomutt] index: Invoking
6626 O C Mar 31 Werner Koch     (4.8K) └─>Re: [neomutt/neomutt] index: Invoking
6627     Mar 30 Amir Goldstein  (2.7K) [PATCH] fanotify: Document FAN_REPORT_FD_
6628     Mar 31 Jan Kara        (3.1K) └─>
6629 r   Mar 31 Amir Goldstein  (2.1K)   └─>
6630  sF Mar 31 To Amir Goldste (2.1K)     └─>
6631   + Mar 31 Mail Delivery S (5.6K) Undelivered Mail Returned to Sender
6632  PF Mar 31 To alx@kernel.o (2.2K) ...




---NeoMutt: ~/mail/migadu/foss/Inbox [Msgs:6632/6632 Old:28 180M]---(threads/las
Invoking PGP...Please enter the passphrase to unlock the OpenPGP secret key:
                                                                            "Alejandro Colomar <alx@alejandro-colomar.es>"
                                          4096-bit RSA key, ID 979EC5380A85F2EC,
                                                                               created 2025-02-14 (main key ID D57633D441E25BB5).

                                                 Passphrase: 

```

which I need to close with `^C`.

**Comment by dd9jn** [comment]
To see more also put a "verbose" intp gpg-agent.conf.  Did you restart the agent?  (gpgconf -K all).

**Comment by alejandro-colomar** [comment]
> To see more also put a "verbose" intp gpg-agent.conf. Did you restart the agent? (gpgconf -K all).

Thanks!  I forgot to kill it.  Here's what I got:

```sh
alx@devuan:~/.gnupg$ ls -l gpg-agent.conf 
-rw------- 1 alx alx 61 Apr  1 12:34 gpg-agent.conf
alx@devuan:~/.gnupg$ cat gpg-agent.conf 
debug-pinentry
log-file /home/alx/.gnupg/alx-dbg.log
verbose
alx@devuan:~/.gnupg$ ls -l /home/alx/.gnupg/alx-dbg.log 
ls: cannot access '/home/alx/.gnupg/alx-dbg.log': No such file or directory
alx@devuan:~/.gnupg$ gpgconf -K all
alx@devuan:~/.gnupg$ ls -l /home/alx/.gnupg/alx-dbg.log 
ls: cannot access '/home/alx/.gnupg/alx-dbg.log': No such file or directory
alx@devuan:~/.gnupg$ xfce4-terminal -e bash
alx@devuan:~/.gnupg$ ls -l /home/alx/.gnupg/alx-dbg.log 
-rw------- 1 alx alx 1457 Apr  1 12:35 /home/alx/.gnupg/alx-dbg.log
```
```sh
alx@devuan:~/.gnupg$ cat !$
cat /home/alx/.gnupg/alx-dbg.log
2025-04-01 12:35:47 gpg-agent[14065] listening on socket '/run/user/1000/gnupg/S.gpg-agent'
2025-04-01 12:35:47 gpg-agent[14065] listening on socket '/run/user/1000/gnupg/S.gpg-agent.extra'
2025-04-01 12:35:47 gpg-agent[14065] listening on socket '/run/user/1000/gnupg/S.gpg-agent.browser'
2025-04-01 12:35:47 gpg-agent[14065] listening on socket '/run/user/1000/gnupg/S.gpg-agent.ssh'
2025-04-01 12:35:47 gpg-agent[14066] gpg-agent (GnuPG) 2.2.46 started
2025-04-01 12:35:47 gpg-agent[14066] handler 0x7f0f1d36c6c0 for fd 10 started
2025-04-01 12:35:47 gpg-agent[14066] starting a new PIN Entry
2025-04-01 12:35:47 gpg-agent[14066] DBG: pinentry: atfork used setenv(DISPLAY,:0.0)
2025-04-01 12:35:47 gpg-agent[14066] DBG: pinentry: atfork used setenv(XAUTHORITY,/home/alx/.Xauthority)
2025-04-01 12:35:47 gpg-agent[14066] DBG: pinentry: atfork used setenv(XDG_SESSION_TYPE,x11)
2025-04-01 12:35:47 gpg-agent[14066] DBG: pinentry: atfork used setenv(DBUS_SESSION_BUS_ADDRESS,unix:path=/tmp/dbus-stGWC0aDTt,guid=974b4da3e38484280e5082a867ebb911)
2025-04-01 12:35:52 gpg-agent[14066] failed to unprotect the secret key: Operation cancelled
2025-04-01 12:35:52 gpg-agent[14066] failed to read the secret key
2025-04-01 12:35:52 gpg-agent[14066] command 'PKDECRYPT' failed: Operation cancelled <Pinentry>
2025-04-01 12:35:52 gpg-agent[14066] Assuan processing failed: Broken pipe
2025-04-01 12:35:52 gpg-agent[14066] handler 0x7f0f1d36c6c0 for fd 10 terminated
```

**Comment by dd9jn** [comment]
Also add "debug ipc" to see also the status lines returned from gpg-agent to gpg and then to gpgme.  Sorry forgot about this.

I assume you canceled on purpose.

**Comment by alejandro-colomar** [comment]
> Also add "debug ipc" to see also the status lines returned from gpg-agent to gpg and then to gpgme. Sorry forgot about this.

Thanks!  No problem.

> 
> I assume you canceled on purpose.

Yep, once I hit this bug, I can't quit neomutt(1) or mutt(1) nicely.  Some keystrokes are read by neomutt(1) and others I assume are read by pinentry-tty(1), so I can't successfully quit.  The only option is to kill the terminal.

**Comment by alejandro-colomar** [comment]
> It might be easier to use pinentry-curses or the standard pinentries, which fallback to curses. The pinentry-tty was written for pure CLI applications.

The thing is, I'd like to have pinentry-tty(1) for most programs, so I want it be the default.

Maybe mutt(1) could request pinentry-curses(1) explicitly somehow?

**Comment by alejandro-colomar** [comment]
> On Tue,  1 Apr 2025 05:28, Alejandro Colomar said:
>  Maybe [neo]mutt(1) could request pinentry-curses(1) explicitly somehow?
> You need to write a wrapper which evaluates the PINENTRY_USER_DATA
> variable to decide which pinentry to call.  Mutt would the set this
> variabale to some specific value.

Nice!  Sounds like it would work.

How does that variable work?  The documentation isn't very clear:

```sh
alx@devuan:/usr/share/man$ find . -type f | xargs zgrep -l PINENTRY_USER_DATA 2>/dev/null;
./man1/gpg.1.gz
alx@devuan:/usr/share/man$ MANWIDTH=72 man gpg | sed -n '/PINENTRY_USER_DATA/,/^$/p'
     PINENTRY_USER_DATA
            This value is passed via gpg‐agent to pinentry.  It is use‐
            ful to convey extra information to a custom pinentry.

```

**Comment by dd9jn** [comment]
On Wed,  2 Apr 2025 02:32, Alejandro Colomar said:

> How does that variable work?  The documentation isn't very clear:

Do whatever you want with it.  It is merely passed along from gpg to
gpg-agent to Pinentry.


Salam-Shalom,

   Werner

-- 
The pioneers of a warless world are the youth that
refuse military service.             - A. Einstein

**Comment by alejandro-colomar** [comment]
> On Wed,  2 Apr 2025 02:32, Alejandro Colomar said:
>  How does that variable work?  The documentation isn't very clear:
> Do whatever you want with it.  It is merely passed along from gpg to
> gpg-agent to Pinentry.

Is it passed as an environment vairbable, as an argument, or how?

**Comment by alejandro-colomar** [comment]
On the other hand, I think it would be safer if gpgme had a way to request a specific pinentry program.  I don't trust myself to write a safe pinentry wrapper.

I think it would be useful to have an equivalent of gpg-agent(1)'s `--pinentry-program` option in gpgme.

**Comment by dd9jn** [comment]
On Wed,  2 Apr 2025 06:16, Alejandro Colomar said:

> On the other hand, I think it would be safer if gpgme had a way to
> request a specific pinentry program.  I don't trust myself to write a

That does not fit weel into the architecture.  It is in particular
problematic because only gpg-agent is responsible for the pinentry.

Create a file /usr/local/bin/mypinentry

--8<---------------cut here---------------start------------->8---
#!/bin/sh
set -e
if [ "$PINENTRY_USER_DATA" = "use-tty" ]; then
  MYPINENTRY=/usr/local/bin/pinentry-tty
else
  MYPINENTRY=/usr/local/bin/pinentry-curses
fi

exec "$MYPINENTRY"
--8<---------------cut here---------------end--------------->8---

chmod +x /usr/local/bin/mypinentry

and put

pinentry-program /usr/local/bin/pinentry-gtk-2

into gpg-agent.conf . Then you only need to start a program like this

PINENTRY_USER_DATA=use-tty myprogram

Or do it vice-versa if you want to have a different default.


Salam-Shalom,

   Werner

-- 
The pioneers of a warless world are the youth that
refuse military service.             - A. Einstein

**Comment by alejandro-colomar** [comment]
@dd9jn Thanks!  This works.

However, I did the opposite: default to tty, and opt-in to curses:

```sh
$ cat ~/.gnupg/gpg-agent.conf 
pinentry-program /usr/local/bin/my-pinentry
```

```sh
$ cat /usr/local/bin/my-pinentry 
#!/bin/sh

set -e;

if test "$PINENTRY_USER_DATA" = "curses"; then
	MY_PINENTRY=/usr/bin/pinentry-curses;
else
	MY_PINENTRY=/usr/bin/pinentry;
fi;

exec "$MY_PINENTRY";
```

```sh
$ which pinentry
/usr/bin/pinentry
$ which pinentry | xargs readlink
/etc/alternatives/pinentry
$ which pinentry | xargs readlink | xargs readlink
/usr/bin/pinentry-tty
```

```sh
$ which pinentry-curses
/usr/bin/pinentry-curses
```

Now I need to run neomutt as

```sh
PINENTRY_USER_DATA=curses neomutt
```

I tried to use the muttrc to do that, but it seems it doesn't work

```diff
diff --git a/etc/Muttrc.d/alx.mutt.rc b/etc/Muttrc.d/alx.mutt.rc
index 67dddfd..e894983 100644
--- a/etc/Muttrc.d/alx.mutt.rc
+++ b/etc/Muttrc.d/alx.mutt.rc
@@ -1,3 +1,4 @@
+source /etc/Muttrc.d/alx/env.muttrc
 source /etc/Muttrc.d/alx/crypt.muttrc
 source /etc/Muttrc.d/alx/index.muttrc
 source /etc/Muttrc.d/alx/pager.muttrc
diff --git a/etc/Muttrc.d/alx/env.muttrc b/etc/Muttrc.d/alx/env.muttrc
new file mode 100644
index 0000000..2c8659b
--- /dev/null
+++ b/etc/Muttrc.d/alx/env.muttrc
@@ -0,0 +1 @@
+setenv PINENTRY_USER_DATA curses
```

@dd9jn , @flatcap  Do you know why?  I see the variable set if I query it in neomutt(1), but it doesn't seem to arrive at my-pinentry.

**Comment by alejandro-colomar** [comment]
> [@dd9jn](https://github.com/dd9jn) , [@flatcap](https://github.com/flatcap) Do you know why? I see the variable set if I query it in neomutt(1), but it doesn't seem to arrive at my-pinentry.

Hmmm, I suspect that since pinentry is not invoked by neomutt(1) itself, but rather through a library, neomutt(1) doesn't set its environment, and since setenv doesn't affect neomutt(1) itself, but rather its children, then the variable isn't set.

@flatcap should `setenv` affect the environment of neomutt(1) itself?  Or is that too complex, and I should rather create an alias.

```sh
alias neomutt='PINENTRY_USER_DATA=curses neomutt'
```

**Comment by flatcap** [comment]
Hmm... NeoMutt keeps a separate *clean* copy of the external environment `NeoMutt->env`.
Your config `setenv PINENTRY_USER_DATA curses` gets added to this.

This is to isolate different parts of the code.
e.g. The mailcap code alters `$COLUMNS` to fit the output into the Pager.

When NeoMutt calls external filters, it passes this Environment to them.
Running, `<shell-escape>bash<enter>`, bash gets `PINENTRY_USER_DATA`.

However, I suspect that using GPGME means that NeoMutt's just making library calls,
so the crypto code is using NeoMutt's **real** environment (sans `PINENTRY_USER_DATA`).

**Comment by alejandro-colomar** [comment]
> However, I suspect that using GPGME means that NeoMutt's just making library calls, so the crypto code is using NeoMutt's **real** environment (sans `PINENTRY_USER_DATA`).

Yep, that's what I suspected.  Are we okay with that?  I don't want to make neomutt(1) too complex, so if you think it's better to not touch this, I'll live with the alias, which is fine to me.

**Comment by flatcap** [comment]
> @flatcap should `setenv` affect the environment of neomutt(1) itself?

Maybe
I haven't checked but I'd imagine this behaviour goes back a long way.

> Or is that too complex

Hmm... dunno.

We'd have to identify all the code that sets environment variables (there's probably not much).
Any config, or manual `:setenv` commands, could be put into the *real* environment.

Then the mailcap code could copy and add to the *real* environment and pass it to the child.
(I'm not sure if there's any other code that manipulates the environment)

I'd like to see an analysis, before a PR ;-)

> and I should rather create an alias?
> alias neomutt='PINENTRY_USER_DATA=curses neomutt'

This will work and requires no code changes :-)

**Comment by alejandro-colomar** [comment]
> > [@flatcap](https://github.com/flatcap) should `setenv` affect the environment of neomutt(1) itself?
> 
> Maybe I haven't checked but I'd imagine the current behaviour goes back a long way.
> 
> > Or is that too complex
> 
> Hmm... dunno.
> 
> We'd have to identify all the code that sets environment variables (there's probably not much). Any config, or manual `:setenv` commands, could be put into the _real_ environment.
> 
> Then the mailcap code could copy and add to the _real_ environment and pass it to the child. (I'm not sure if there's any other code that manipulates the environment)
> 
> I'd like to see an analysis, before a PR ;-)

Yeah, and I'm also worried that it might have some unintended consequences, so let's not do it (unless someone feels like doing the job of making sure it's fine).  :-)

> 
> > and I should rather create an alias?
> > alias neomutt='PINENTRY_USER_DATA=curses neomutt'
> 
> This will work and requires no code changes :-)

Yep, I'll do this.  So, I'll close this issue finally!  Thank you all!  :-)
